### PR TITLE
LPS-113200 prove Management Toolbar Creation Menu "More" button behavior

### PR DESCRIPTION
**Pull request by wincent**
Forwarding you a copy of this from https://github.com/wincent/liferay-portal/pull/305 because I am still battling with Commerce issues in https://github.com/wincent/liferay-portal-ee/pull/9 and it is possible you will be able to look at it before I get to it, and in any case, my capacity to usefully review is pretty eroded at this point.

Original message follows.

--- 

Hey @wincent ,

This is a followup on https://github.com/jbalsas/liferay-portal/pull/2172. We are updating remaining modals in `journal-web` and `document-library-web`.

The problem with the 'more' modal is explained in detail in https://issues.liferay.com/browse/LPS-113200. The goal in this PR is to ensure the default menu behavior, extension, happens only if modal is not implemented:
1. We are wrapping default behavior in `setTimeout(extendMenu, 0)`, ensuring that handler defined in modules executes first.
1. This is enough for `journal-web`, but in `document-library-web` we are opening `Liferay.Util.openModal` from AUI module, meaning it opens asynchronously. That is why we are adding `Liferay.creationMenuMoreModal` and synchronizing over it.

So, the order of execution is:
1. Handler in `frontend-taglib-clay`. We do nothing, delay its execution.
1. Handler in `document-library-web`. We set `Liferay.creationMenuMoreModal`
1. Delayed execution in `frontend-taglib-clay`. If `Liferay.creationMenuMoreModal` was set, we prevent default behavior.
1. Modal opens.

---
Test in document library:
1. Documents and Media > Doyument Types tab > Create 10 document types
1. Documents and Media tab > (+) > More
1. Close modal.
1. Documents and Media tab > (+) > More
1. Select a type. This opens a new input page.
---
Test in journal:
1. Web Content > Structures > Create 7 structures
1. Web Content tab > (+) > More
1. Close modal.
1. Web Content tab > (+) > More
1. Select a structure. This opens a new input page.
---
Test of default behavior:
1. Site Builder > Widget Templates > (+) > More 

Please review the code.

Thank you!

/cc @jbalsas @carloslancha

### CODE REVIEW

**Comment by jbalsas** [comment]
Hey @markocikos, I'm trying to take a look here today!

> Test of default behavior:
>
> Site Builder > Widget Templates > (+) > More

Well... this is actually a escape hatch, as noted in the [Lexicon Docs](https://liferay.design/lexicon/core-components/dropdowns/)

> Note: The default behaviour of the More button is to load more items in the dropdown menu, but it is not the desired one.
>
> [...]
>
> Include a More area to give access to the modal that contains the rest of the items and lets the user customize the list of favorites.

@drakonux... do these affirmations still hold true?

We just implemented that as a fallback, so that those applications not implementing the modal wouldn't get a useless button and a cropped list...

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/ManagementToolbar.es.js [hunk 1]
```diff
@@ -135,34 +135,42 @@ class ManagementToolbar extends ClayComponent {
...
-		this.creationMenu.maxPrimaryItems = creationMenuPrimaryItemsCount;
-		this.creationMenu.maxSecondaryItems = creationMenuSecondaryItemsCount;
-		this.creationMenu.maxTotalItems = creationMenuTotalItemsCount;
-
-		this.refs.managementToolbar.refs.creationMenuDropdown.maxPrimaryItems = creationMenuPrimaryItemsCount;
-		this.refs.managementToolbar.refs.creationMenuDropdown.maxSecondaryItems = creationMenuSecondaryItemsCount;
-		this.refs.managementToolbar.refs.creationMenuDropdown.maxTotalItems = creationMenuTotalItemsCount;
+	_handleCreationMenuMoreButtonClicked(event) {
+		const extendMenu = () => {
+			if (Liferay.creationMenuMoreModal) {
```
Comment (jbalsas) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/management_toolbar/ManagementToolbar.es.js:140]: Just do this instead:

```javascript
if (event.preventedDefault) {
	return;
}
```

**Comment by jbalsas** [comment]
Hey @markocikos, I did a quick check, and since you're already microtasking the extendMenu call, you could simply check to see if the event has been prevented to avoid expanding the elements in the list.

Can you push that change and resend? Should be good to go!

**Comment by drakonux** [comment]
The "More" button is a fallback as you said. However the common behavior in a product is a dropdown menu with a fixed number of items and that's it.

Said this, there are scenarios that allow users to create dynamically many items. So we should provide flexibility allowing a product team to use just a fixed height in the dropdown menu and when the items overflow, the vertical scrollbar should be shown. Or a combination of both, fixed height and the "More" button to load the rest of the items triggering the overflow behavior of the list or shown a modal window to customize the list.

**Comment by markocikos** [comment]
Followup in https://github.com/jbalsas/liferay-portal/pull/2202

**Comment by markocikos** [comment]
ci:close

**Pull request closed without merging**